core/src/main/java/org/apache/iceberg/TableMetadataParser.java

```diff
@@ -122,15 +122,25 @@ public static void write(TableMetadata metadata, OutputFile outputFile) {
   public static void internalWrite(
       TableMetadata metadata, OutputFile outputFile, boolean overwrite) {
     boolean isGzip = Codec.fromFileName(outputFile.location()) == Codec.GZIP;
-    try (OutputStream os = overwrite ? outputFile.createOrOverwrite() : outputFile.create();
-        OutputStream gos = isGzip ? new GZIPOutputStream(os) : os;
+    OutputStream os = overwrite ? outputFile.createOrOverwrite() : outputFile.create();
+    // if isGzip is true, os will be closed by GZIPOutputStream,
+    // otherwise, os will be closed by try-with-resources
+    try (OutputStream gos = isGzip ? new GZIPOutputStream(os) : os;
         OutputStreamWriter writer = new OutputStreamWriter(gos, StandardCharsets.UTF_8)) {
       JsonGenerator generator = JsonUtil.factory().createGenerator(writer);
       generator.useDefaultPrettyPrinter();
       toJson(metadata, generator);
       generator.flush();
     } catch (IOException e) {
       throw new RuntimeIOException(e, "Failed to write json to file: %s", outputFile);
+    } finally {
+      try {
+        // in case of an exception in GZIPOutputStream constructor,
+        // the stream is not closed and needs to be closed here
```

Comment:
"Explicitly close the stream in case of exceptions thrown by GZIPOutputStream constructor. See https://github.com/apache/iceberg/pull/11220 for more details"